In [173]:
import json
import numpy as np
import sys

In [174]:
plan = [{'days': '1',
  'breakfast': 'TwnzM8mJn_nT2PJf1x-9kQ',
  'morning_attractions': ['Ib9HV7ekw459jM1Ksdiyiw'],
  'lunch': 'wHkYLlZyPXllrQRlvidUlg',
  'afternoon_attractions': ['kbEVlzQLcYS3JSQPG9QMOQ',
   '8_O6LXLyMgpq1g9CIwcW4w'],
  'dinner': 'S8ZFYEgMejpChID8tzKo9A',
  'night_attractions': ['pxZAz8pv18wK_t-m8WpN0g'],
  'accommodation': '3QPAh9VvYNTnqAFgBeBcng'},
 {'days': '2',
  'breakfast': 'rYqmaOIULRouz_1db07OdQ',
  'morning_attractions': ['Qw7tz-UkPrpXaVidWuab4Q'],
  'lunch': '6ajnOk0GcY9xbb5Ocaw8Gw',
  'afternoon_attractions': ['4mWzXhD8vo0bABVCGAhlqA',
   'ytynqOUb3hjKeJfRj5Tshw'],
  'dinner': '05ev984NYfimRN0UiFrxaA',
  'night_attractions': ['dBRWKIS7h-qZCi8EjUJ2HA'],
  'accommodation': '3QPAh9VvYNTnqAFgBeBcng'}] 

with open('../Datasets/Attractions_task1.jsonl', 'r') as f:
    data = [json.loads(line) for line in f]

with open('../Datasets/Hotels_task1.jsonl', 'r') as f:
    data_hotel = [json.loads(line) for line in f]

def getCordinate(id):
    for attraction in data:
        if attraction['business_id'] == id:
            return (attraction['latitude'], attraction['longitude'])
def getCordinate_Hotel(id):
    for hotel in data_hotel:
        if hotel['business_id'] == id:
            return (hotel['latitude'], hotel['longitude'])

In [175]:
#get the cordinates
cordinates = []
for day in plan:
    cordinate_one_day = []
    
    cordinate_one_day.append(getCordinate_Hotel(day['accommodation']))
    
    for attraction in day['morning_attractions']:
        cordinate_one_day.append(getCordinate(attraction))
    for attraction in day['afternoon_attractions']:
        cordinate_one_day.append(getCordinate(attraction))
    for attraction in day['night_attractions']:
        cordinate_one_day.append(getCordinate(attraction))
            
    cordinates.append(cordinate_one_day)
cordinates

[[(39.946396818, -75.1609669488),
  (39.9488980012, -75.1500296367),
  (39.9495774962, -75.1503095688),
  (39.953893, -75.149068),
  (39.9445424967, -75.1414754636)],
 [(39.946396818, -75.1609669488),
  (39.965573, -75.180969),
  (39.942712, -75.159313),
  (39.9533414645, -75.1588545174),
  (39.9582109, -75.1731373822)]]

In [176]:
def getDistanceMatrix(cordinates):
    n = len(cordinates)
    distance_matrix = np.zeros((n, n))
    for i in range(n):
        for j in range(i+1, n):
            distance_matrix[i][j] = distance_matrix[j][i] = ((cordinates[i][0]*1000 - cordinates[j][0]*1000)**2 + (cordinates[i][1]*1000 - cordinates[j][1]*1000)**2)**0.5
    return distance_matrix

In [177]:
cordinates_one_day = cordinates[0]

distance_matrix = getDistanceMatrix(cordinates_one_day)
distance_matrix

array([[ 0.        , 11.21965745, 11.12189113, 14.06334694, 19.57949189],
       [11.21965745,  0.        ,  0.73489825,  5.08672372,  9.59918209],
       [11.12189113,  0.73489825,  0.        ,  4.49055299, 10.16821689],
       [14.06334694,  5.08672372,  4.49055299,  0.        , 12.04485454],
       [19.57949189,  9.59918209, 10.16821689, 12.04485454,  0.        ]])

In [178]:
def totalCost(mask, pos, n, cost, info_lists):
    distance_list = []
    i_list = []
    # Base case: if all cities are visited, return the
    # cost to return to the starting city (0)
    if mask == (1 << n) - 1:
        return cost[pos][0]

    ans = sys.maxsize   

    # Try visiting every city that has not been visited yet
    for i in range(n):
        if (mask & (1 << i)) == 0: 
            i_list.append(i)
            # If city i is not visited, visit it and 
             #  update the mask
            distance_list.append(cost[pos][i] +
                      totalCost(mask | (1 << i), i, n, cost, info_lists))
        

    info_list = [i_list, distance_list]
    info_lists.append(info_list)
    
    ans = min(distance_list)
    return ans

In [179]:
n = len(distance_matrix)
info_list = []
optimized_distance = totalCost(1, 0, n, distance_matrix,info_list)
print("the short distance is: ", optimized_distance)

the short distance is:  47.564172952389654


In [180]:
#get a list of 1 to n
candidates = list(range(n-1))
#add 1 to the values
candidates = [x+1 for x in candidates]

moves = []

while len(candidates) > 0:
    #find the last one in the lnfo_list
    for i in range(len(info_list)):
        if info_list[i][0] == candidates:
            next_move = info_list[i][0][np.argmin(info_list[i][1])]
            moves.append(next_move)
            #take next move out of candidates
            candidates.remove(next_move)

moves_reversed = moves[::-1]
optimized_route = [[0] + moves, [0] + moves_reversed]
print(" the shortest path is: ", optimized_route)
            

 the shortest path is:  [[0, 2, 1, 3, 4], [0, 4, 3, 1, 2]]


Validation!

In [181]:
distance_matrix = [
        [0, 10, 15, 50],
        [10, 0, 50, 25],
        [15, 50, 0, 30],
        [50, 25, 30, 0]
    ]

In [182]:
n = len(distance_matrix)
info_list = []
optimized_distance = totalCost(1, 0, n, distance_matrix,info_list)
print(optimized_distance)

80


In [183]:
info_list

[[[3], [80]],
 [[2], [45]],
 [[2, 3], [130, 70]],
 [[3], [75]],
 [[1], [35]],
 [[1, 3], [125, 65]],
 [[2], [65]],
 [[1], [60]],
 [[1, 2], [90, 90]],
 [[1, 2, 3], [80, 80, 140]]]

In [184]:
#get a list of 1 to n
candidates = list(range(n-1))
#add 1 to the values
candidates = [x+1 for x in candidates]

moves = []

while len(candidates) > 0:
    #find the last one in the lnfo_list
    for i in range(len(info_list)):
        if info_list[i][0] == candidates:
            next_move = info_list[i][0][np.argmin(info_list[i][1])]
            moves.append(next_move)
            #take next move out of candidates
            candidates.remove(next_move)

moves_reversed = moves[::-1]
optimized_route = [[0] + moves + [0], [0] + moves_reversed + [0]]
print(optimized_route)

[[0, 1, 3, 2, 0], [0, 2, 3, 1, 0]]


Multi day

In [185]:
def totalCost_multiday(mask, pos, day, n, cost, info_lists):
    current_day_visits = len(cordinates[day])
    distance_list = []
    i_list = []
    # Base case: if all cities are visited, return the
    # cost to return to the starting city (0)
    if mask == (1 << n) - 1:
        return cost[pos][0]

    ans = sys.maxsize   

    # Try visiting every city that has not been visited yet
    for i in range(n):
        if (mask & (1 << i)) == 0: 
            i_list.append(i)
            # If city i is not visited, visit it and 
             #  update the mask
            distance_list.append(cost[pos][i] +
                      totalCost(mask | (1 << i), i, day, n, cost, info_lists))
        

    info_list = [i_list, distance_list]
    info_lists.append(info_list)
    
    ans = min(distance_list)
    return ans

In [186]:
cordinates

[[(39.946396818, -75.1609669488),
  (39.9488980012, -75.1500296367),
  (39.9495774962, -75.1503095688),
  (39.953893, -75.149068),
  (39.9445424967, -75.1414754636)],
 [(39.946396818, -75.1609669488),
  (39.965573, -75.180969),
  (39.942712, -75.159313),
  (39.9533414645, -75.1588545174),
  (39.9582109, -75.1731373822)]]

In [187]:
cordinates[0][0]

(39.946396818, -75.1609669488)

In [188]:
cordinates_all_days = []
for i in range(len(cordinates)):
    for business in cordinates[i]:
        cordinates_all_days.append(business)

cordinates_all_days

[(39.946396818, -75.1609669488),
 (39.9488980012, -75.1500296367),
 (39.9495774962, -75.1503095688),
 (39.953893, -75.149068),
 (39.9445424967, -75.1414754636),
 (39.946396818, -75.1609669488),
 (39.965573, -75.180969),
 (39.942712, -75.159313),
 (39.9533414645, -75.1588545174),
 (39.9582109, -75.1731373822)]

In [189]:
distance_matrix = getDistanceMatrix(cordinates_all_days)
distance_matrix

array([[ 0.        , 11.21965745, 11.12189113, 14.06334694, 19.57949189,
         0.        , 27.70934875,  4.03898878,  7.25882094, 16.96148527],
       [11.21965745,  0.        ,  0.73489825,  5.08672372,  9.59918209,
        11.21965745, 35.14683181, 11.15560151,  9.88042942, 24.91381115],
       [11.12189113,  0.73489825,  0.        ,  4.49055299, 10.16821689,
        11.12189113, 34.58116342, 11.32240308,  9.33721607, 24.40583385],
       [14.06334694,  5.08672372,  4.49055299,  0.        , 12.04485454,
        14.06334694, 33.97199142, 15.16491958,  9.80204643, 24.45361773],
       [19.57949189,  9.59918209, 10.16821689, 12.04485454,  0.        ,
        19.57949189, 44.74395475, 17.93121365, 19.47956225, 34.48626304],
       [ 0.        , 11.21965745, 11.12189113, 14.06334694, 19.57949189,
         0.        , 27.70934875,  4.03898878,  7.25882094, 16.96148527],
       [27.70934875, 35.14683181, 34.58116342, 33.97199142, 44.74395475,
        27.70934875,  0.        , 31.48980243

In [190]:
import pandas as pd

# Convert the array into a DataFrame (table)
df = pd.DataFrame(distance_matrix)

# Save the DataFrame to an Excel file
df.to_excel('output.xlsx', index=False)

In [191]:
def getHotelIndex(day,cordinates):
    hotel_index = 0
    if day > 0:
        for j in range(day):
            hotel_index += len(cordinates[j])
    return hotel_index

def totalCost_multiday(mask, pos, day, cordinates, n, visited, cost, info_lists):
    visit_requirement = len(cordinates[day])
    distance_list = []
    i_list = []

    hotel_index = getHotelIndex(day,cordinates)
    hotel_index_next_day = getHotelIndex(day+1,cordinates)
    # Base case: if all cities are visited, return the
    # cost to return to the starting city (0)

    if mask == (1 << n) - 1:
        return cost[pos][hotel_index]

    if visit_requirement == visited:
        for i in range(n):
            if (mask & (1 << i)) == 0: 
                i_list.append(i)
                distance_list.append(cost[pos][i] +totalCost_multiday(mask | (1 << i), hotel_index_next_day, day + 1, cordinates, n, 1, cost, info_lists))
        
        info_list = [i_list, distance_list]
        info_lists.append(info_list)
        
        return min(distance_list) + cost[pos][hotel_index] # change this to the old hotel position
    
    # Try visiting every city that has not been visited yet
    for i in range(n):
        if (mask & (1 << i)) == 0: 

            i_list.append(i)
            # If city i is not visited, visit it and 
             #  update the mask
            distance_list.append(cost[pos][i] +
                      totalCost_multiday(mask | (1 << i), i, day, cordinates, n, visited + 1, cost, info_lists))
        

    info_list = [i_list, distance_list]
    info_lists.append(info_list)
  
    return min(distance_list)

In [192]:
n = len(distance_matrix)
info_list = []
#newMask will have all the hotels as 1 before the function.
newMask = 1
index_list = list(range(len(cordinates) - 1))
index_list = index_list[::-1]
newMask = 1
for i in index_list:
    newMask = (newMask << (len(cordinates[i]))) + 1

optimized_distance = totalCost_multiday(newMask, 0,0,cordinates,n,1,distance_matrix, info_list)

In [193]:
n

10

In [194]:
newMask

33

In [195]:
optimized_distance

np.float64(105.48895393859405)

This might be correct! The seperate ones are 47 + 67, and pretty optimized, but if the new number 105 seems pretty correct.